server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java

```diff
@@ -0,0 +1,251 @@
+        final Repository repository = repositoriesService.repository(repositoryName);
+        if (repository instanceof BlobStoreRepository == false) {
+            listener.onFailure(new IllegalArgumentException("Repository [" + repositoryName + "] does not support repository cleanup"));
+            return;
+        }
+        final long repositoryStateId = repository.getRepositoryData().getGenId();
+        logger.info("Running cleanup operations on repository [{}][{}]", repositoryName, repositoryStateId);
+        clusterService.submitStateUpdateTask("cleanup repository", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                SnapshotDeletionsInProgress deletionsInProgress = currentState.custom(SnapshotDeletionsInProgress.TYPE);
+                final RepositoryCleanupInProgress repositoryCleanupInProgress = currentState.custom(RepositoryCleanupInProgress.TYPE);
+                if (repositoryCleanupInProgress != null && repositoryCleanupInProgress.cleanupInProgress() == false) {
+                    throw new IllegalStateException(
+                        "Cannot cleanup [" + repositoryName + "] - a repository cleanup is already in-progress");
+                }
+                if (deletionsInProgress != null && deletionsInProgress.hasDeletionsInProgress()) {
+                    throw new IllegalStateException("Cannot cleanup [" + repositoryName + "] - a snapshot is currently being deleted");
+                }
+                ClusterState.Builder clusterStateBuilder = ClusterState.builder(currentState);
+                SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
+                if (snapshots != null && !snapshots.entries().isEmpty()) {
+                    throw new IllegalStateException("Cannot cleanup [" + repositoryName + "] - a snapshot is currently running");
+                }
+                final RepositoryCleanupInProgress newCleanupInProgress =
+                    new RepositoryCleanupInProgress(RepositoryCleanupInProgress.startedEntry(repositoryName, repositoryStateId));
+                return clusterStateBuilder.putCustom(RepositoryCleanupInProgress.TYPE, newCleanupInProgress).build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                after(e, null);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                logger.debug("Initialized repository cleanup in cluster state for [{}][{}]", repositoryName, repositoryStateId);
+                threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new ActionRunnable<>(listener) {
+                    @Override
+                    protected void doRun() {
+                        Repository repository = repositoriesService.repository(repositoryName);
```

Comment:
You raise a good point here :)

The cleanup operation is a little less elegant than the snapshot create operation when it comes to handling the repository state id. We read the state id up-front and then create the entry in the CS. So if we have some concurrency here (easily possible) where a snapshot just finishes creating and we still read the old index gen, then we're in the unfortunate situation where the cleanup will fail even though it was set up in the CS and even if there was no master fail-over. 
Just reading again here won't do us much good (compared to what already happens), because if we now see a changed repository state id here we would still have to submit another state update task to align the repo state IDs.

That said: We have the same issue with snapshot deletion as well => I'm wondering if we should just live with it here to keep things simple and then in a follow-up clean up both by adding some notion of `INIT` state to them like we do for snapshot create (so we "lock" the repo before reading the state id)?

Reply:
Actually, on second thought I think this is fine. No one ever complained about the equivalent and incredibly unlikely situation for snapshot delete and create either ... We can add stages to the cleanup once we get to deeper cleanup on a shard level imo.

Reply:
I looked at various scenarios and I agree with your reasoning. I also think that reading again from cluster state won't be better after all. 

I think this is fine. On the question of adding some INIT stage to cluster state object markers I need to think about it again.